File: SnS/TermProject/Sift/ImPreprocessing.py
# import math
import numpy

def gaussian_blur(input, sigma=1.6, size=-1):
    """
    Gaussian blurring using 2-dimensional square Gaussian kernel.

    :param input: Image or ndarray
        Input image.
    :param sigma: scalar (sequence of scalars are not supported yet)
        Standard deviation for Gaussian kernel. According to Lowe in his
        famous SIFT paper, sigma is set to be 1.6 (default here).
    :param size: scalar (sequence of scalars are not supported yet)
        Size of the Gaussian kernel. If being negative or zero, size will
        be 'around' (6*sigma+1).
        size should be odd. If not, it will be increased by 1.
    :return: ndarray
        Returned array of same shape as 'input'.

    """

    input = numpy.asanyarray(input)

    if size <= 0:
        size = int(numpy.ceil(6 * sigma + 1))
    if size % 2 == 0:
        size += 1
    gaussian_array1d = numpy.arange(-(size - 1) / 2, (size - 1) / 2 + 1)
    gaussian_array1d = numpy.exp(-(gaussian_array1d ** 2) / (2 * sigma)) / \
                       (2 * numpy.pi * (sigma ** 2))
    gaussian_array1d /= sum(gaussian_array1d)

    im_conv = numpy.zeros(input.shape)
    imrows, imcols = input.shape
    print "image size: ", input.shape

    for row in range(0, imrows):
        for col in range(0, imcols):
            for index in range(max(-(size - 1) / 2 + col, 0),
                               min((size - 1) / 2 + col + 1, imcols)):
                im_conv[row, col] \
                    += input[row, index] * gaussian_array1d[index - col + (size - 1) / 2]
    for row in range(0, imrows):
        for col in range(0, imcols):
            for index in range(max(-(size - 1) / 2 + row, 0),
                               min((size - 1) / 2 + row + 1, imrows)):
                im_conv[row, col] \
                    += input[index, col] * gaussian_array1d[index - row + (size - 1) / 2]

    return im_conv


if __name__ == "__main__":
    import os
    from PIL import Image
    import scipy
    from scipy import *
    import scipy.misc
    from pylab import *
    import time

    # blank = [[255 for i in range(10)] for j in range(10)]
    # Image.fromarray(uint8(array(blank))).save("./fig/white.jpg")

    testfile = "./fig/Mercedes.jpg"
    test_out_path = "./test_out/"
    if os.path.exists(test_out_path) == False:
        os.makedirs(test_out_path)

    im = array(Image.open(testfile).convert("L"), "f")
    Image.fromarray(uint8(im)).save(
        test_out_path + os.path.splitext(os.path.split(testfile)[1])[0] + "_grayscale.jpg")

    t_start = time.time()
    im_blurred = Image.fromarray(gaussian_blur(im, 100))
    t_finish = time.time()

    print "Gaussian blurring has cost ", t_finish - t_start, " seconds"
    imshow(Image.fromarray(uint8(im_blurred)))
    show()
    Image.fromarray(uint8(im_blurred)).save(
        test_out_path + os.path.splitext(os.path.split(testfile)[1])[0] + "_out.jpg")

File: C&C++/ImagePreprocessing/ImPreprocessing.cpp
#include <python2.7/Python.h>
#include <numpy/arrayobject.h>
#include <stdio.h>
#include <math.h>

/* Gaussian blurring using 2-dimensional square Gaussian kernel. */
double **GaussianBlur(int **input, int rowsOfInput, int colsOfInput,
                      double sigma = 1.6, int size = -1) {
    if (size <= 0)
        size = (int) ceil(6 * sigma + 1);
    if (size % 2 == 0)
        size++;
    double *gaussian_array1d = new double[size];
    int halfSideLength = (size - 1) / 2;
    double sum = 0;
    for (int i = 0; i < size; i++) {
        // the central point of gaussian array lies at i = (size - 1) / 2
        gaussian_array1d[i] = exp(-pow((i - halfSideLength), 2) / (2 * pow(sigma, 2)))
                              / (sqrt(2 * M_PI) * sigma);
        sum += gaussian_array1d[i];
    }
    printf("image size: (%d, %d)\n", rowsOfInput, colsOfInput);

    double **im_conv = new double *[rowsOfInput];
    for (int row = 0; row < rowsOfInput; row++) {
        im_conv[row] = new double[colsOfInput];
        for (int col = 0; col < colsOfInput; col++)
            for (int index = (int) fmax(-(size - 1) / 2 + col, 0);
                 index < fmin((size - 1) / 2 + col + 1, colsOfInput); index++)
                im_conv[row][col]
                        += input[row][index] * gaussian_array1d[index - col + (size - 1) / 2];
    }

    double **im_blurred = new double *[rowsOfInput];
    for (int row = 0; row < rowsOfInput; row++) {
        im_blurred[row] = new double[colsOfInput];
        for (int col = 0; col < colsOfInput; col++)
            for (int index = (int) fmax(-(size - 1) / 2 + row, 0);
                 index < fmin((size - 1) / 2 + row + 1, rowsOfInput); index++)
                im_blurred[row][col]
                        += im_conv[index][col] * gaussian_array1d[index - row + (size - 1) / 2];
    }

    for (int row = 0; row < rowsOfInput; row++)
        delete im_conv[row];
    delete im_conv;
    delete gaussian_array1d;

    return im_blurred;
}


static PyObject *_gaussian_blur(PyObject *self, PyObject *args) {
    int **_input;
    int _rowsOfInput, _colsOfInput;
    int _sigma, _size;
    double **im_blurred;

    if (!PyArg_Parse(args, "O!(ii)di", &PyArray_Type, &_input, &_rowsOfInput,
                     &_colsOfInput, &_sigma, &_size))
        return NULL;

    im_blurred = GaussianBlur(_input, _rowsOfInput, _colsOfInput, _sigma, _size);

    npy_intp dims[_rowsOfInput];
    for (int i = 0; i < _rowsOfInput; i++)
        dims[i] = _colsOfInput;

    PyObject *out = PyArray_SimpleNewFromData(_rowsOfInput, dims,
                                              NPY_DOUBLE, im_blurred);
}

static PyMethodDef ImPreprocModuleMethods[] = {
        {
                "gaussian_blur",
                _gaussian_blur,
                METH_VARARGS,
                ""
        },
        {NULL, NULL, 0, NULL}
};

PyMODINIT_FUNC initImPreprocessing(void) {
    (void) Py_InitModule("ImPreprocessing", ImPreprocModuleMethods);
}